_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+
+.Architecture of platform for running Terraform/Terragrunt CI/CD workflows.
+image::/assets/img/guides/infrastructure-cicd-pipeline/tftg-pipeline-architecture.png[]
+
+.Sequence diagram of running Terraform/Terragrunt CI/CD workflows.
+image::/assets/img/guides/infrastructure-cicd-pipeline/tftg-pipeline-sequence-diagram.png[]
+
+
+[[use_generic_cicd_platforms_as_a_workflow_engine_but_run_infrastructure_deployments_from_within_your_account]]
+=== Use generic CI/CD platforms as a workflow engine but run infrastructure deployments from within your account
+
+As we covered in the <<cicd_platforms>> section of the Core Concepts, all CI/CD platforms support some form of
+integration with VCS repositories. Additionally, CI/CD platforms are optimized for implementing workflows and pipelines
+that react to VCS activities. They all support a way to start executing a pipeline on commits, approval steps in the
+pipeline, and a way to report back to the VCS repository by annotating commits with status of a pipeline execution.
+With the abundance of CI/CD platforms out there, you are bound to find one that suits your needs. It doesn't matter
+which one you use, but be sure to choose one that has enough support for your application code so that you can use the
+same platform for both application and infrastructure CI/CD workflows. In addition, we recommend selecting a Saas
+based CI/CD platform to avoid the overhead of managing additional infrastructure for the VCS integration and workflow
+execution.
+
+However, instead of relying on the CI/CD platform to run infrastructure deployments (e.g calling `terraform apply`), use
+a specialized self-hosted deployment platform for running your infrastructure deployments. This can be Terraform
+Enterprise if you are only using `terraform`, or Gruntwork's ECS Deploy Runner stack for a more general purpose
+platform. In order to support arbitrary infrastructure deployments, it will be inevitable that the system running your
+deployments will have de factor admin status to the environments that you are deploying into. As such, it is important
+that these systems are locked down. You do not want any third party publicly facing system to hold powerful credentials
+to your cloud environments.
+
+This design implements separates out the concerns so that we take full advantage of the strengths of each platform,
+while covering the weaknesses: relying on the CI/CD platforms to manage the workflow/pipline, but have it trigger
+infrastructure deployments on self-hosted systems that are more locked down. In this model, the CI/CD platforms only
+need enough permissions to trigger a deployment, but not necessarily all the permissions required for running a
+deployment. Instead, you can have a system that is closed off in your isolated network hold the credentials
+necessary to actually run the deployment.
+
+
+[[use_serverless_platforms_for_infrastructure_deployments]]
+=== Use serverless platforms for infrastructure deployments
```

Comment:
I don't think serverless is essential here. The important part is that it's elastic enough to handle your needs and thoroughly locked down. E.g., In the future, we should probably allow our CI / CD Fargate task to instead run on K8S (e.g., in an isolated namespace).

Reply:
Replaced with commentary on deploy server options, comparing ECS Deploy Runner with TFE: 427a56bc1a9eb400331fdb34cf82e44a9d6aabcb